### ISSUE TITLE
Windows build broken due to lit

### ISSUE THREAD

**Issue by a1880** [msoos/cryptominisat#531]
Hi Mate,
I've noticed today that the AppVeyor Windows build script breaks after the line "lit --version". 
To try it, I downloaded lit on my system. If I am not mistaken, the version query has to be started
as

python -m lit.main --version

This is what I get as response:
c:\AK\Python36\Lib\runpy.py:125: RuntimeWarning: 'lit.main' found in sys.modules after import of package 'lit', but prior to executi
  warn(RuntimeWarning(msg))
lit 0.7.1

It is possible to start python scripts directly as Windows commands without adding the .py extension. To do that, .py hast to be added to the PATHEXT variable. 

Greetings!

Axel

**Comment by a1880** [comment]
Sadly, the Windows build still does not work. The Visual Studio 2015 build breaks due to further python lit inconsistencies. The Visual Studio 2017 build does not find the boost library program_options. I hope that I'll be able to have a deeper look at this. It would really be nice to have an up-to-date Cryptomininsat executable for Windows.

Greetings and special thanks for your long-standing "Windows patience",

Axel

**Comment by msoos** [comment]
Hi,

Actually, currently it dies on the python "setup.py" not finding the DLL that has just been built:

```
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\BIN\x86_amd64\link.exe /nologo /INCREMENTAL:NO /LTCG /DLL /MANIFEST:EMBED,ID=2 /MANIFESTUAC:NO /LIBPATH:. /LIBPATH:C:/projects/cryptominisat/build/lib /LIBPATH:C:\Python36-x64\libs /LIBPATH:C:\Python36-x64\PCbuild\amd64 "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\ATLMFC\LIB\amd64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\lib\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.14393.0\um\x64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib\amd64" "/LIBPATH:C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\atlmfc\lib\amd64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\10\lib\10.0.10240.0\ucrt\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\8.1\lib\winv6.3\um\x64" "/LIBPATH:C:\Program Files (x86)\Windows Kits\NETFXSDK\4.6.1\Lib\um\x64" cryptominisat5win.lib /EXPORT:PyInit_pycryptosat build\temp.win-amd64-3.6\Release\projects/cryptominisat/build/pycryptosat/src/pycryptosat.obj /OUT:C:\projects\cryptominisat\build\pycryptosat\pycryptosat.cp36-win_amd64.pyd /IMPLIB:build\temp.win-amd64-3.6\Release\projects/cryptominisat/build/pycryptosat/src\pycryptosat.cp36-win_amd64.lib
LINK : fatal error LNK1181: cannot open input file 'cryptominisat5win.lib' [C:\projects\cryptominisat\build\pycryptosat\python_interface.vcxproj]
CUSTOMBUILD : error : command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\x86_amd64\\link.exe' failed with exit status 1181 [C:\projects\cryptominisat\build\pycryptosat\python_interface.vcxproj]
```

See the end of https://ci.appveyor.com/project/msoos/cryptominisat/builds/21130284/job/6pyjwan14lgbcffq for the above issue.

**Comment by msoos** [comment]
It seem that the .lib was put into:
```
Creating library C:/projects/cryptominisat/build/lib/Release/cryptominisat5win.lib
and object C:/projects/cryptominisat/build/lib/Release/cryptominisat5win.exp
```
but it was searching only in:  `/LIBPATH:C:/projects/cryptominisat/build/lib`

**Comment by a1880** [comment]
Hi Mate,
please find below a recipe which I have derived from your installation notes to do a CMS build under Windows with Visual Studio 2017. There are some small differences compared to your notes.

I could build CMS without major problems, but the subsequent pycryptosat part failed with a library search issue I couldn't resolve yet. This is probably related to your comment above 2 days ago.

Greetings,

Axel


Cryptominisat Installation Recipe for Windows + Visual Studio 2017

```
C:\> [ download cryptominisat-version.zip ]  
Example URL for source code release 5.6.6: https://github.com/msoos/cryptominisat/archive/5.6.6.zip

C:\> unzip cryptominisat-version.zip
Use Windows Explorer or tools like 7z. Unzip is not available in Windows by default.
See http://gnuwin32.sourceforge.net/downlinks/unzip-bin-zip.php

C:\> rename cryptominisat-version cms
C:\> cd cms
C:\cms> mkdir build
C:\cms> cd build

C:\cms\build> [ download https://dl.bintray.com/boostorg/release/1.69.0/source/boost_1_69_0.zip ]
see https://studiofreya.com/2017/04/23/building-boost-1-64-with-visual-studio-2017/

C:\cms\build> unzip boost_1_69_0.zip
C:\cms\build> mkdir boost_1_69_0_install
C:\cms\build> cd boost_1_69_0
C:\cms\build\boost_1_69_0> bootstrap.bat --with-libraries=program_options
C:\cms\build\boost_1_69_0> b2 --with-program_options address-model=64 toolset=msvc-14.1 variant=release link=static threading=multi runtime-link=static install --prefix="C:\cms\build\boost_1_69_0_install" > boost_install.out
Check file boost_install.out for irregularities

C:\cms\build\boost_1_69_0> cd ..

C:\cms\build> [ download https://github.com/madler/zlib/archive/v1.2.11.zip ; unzip to zlib sub-directory]
              or: git clone https://github.com/madler/zlib
                  cd zlib
                  git checkout v1.2.8
C:\cms\build> cd zlib
C:\cms\build\zlib> mkdir build
C:\cms\build\zlib> mkdir myinstall
C:\cms\build\zlib> cd build
C:\cms\build\zlib\build> [add full path]cmake -G "Visual Studio 15 2017 Win64" -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=C:\cms\build\zlib\myinstall ..
C:\cms\build\zlib\build> set MSB=E:\Program Files (x86)\Microsoft Visual Studio\2017\Community\MSBuild\15.0\Bin\amd64\MSBuild.exe
C:\cms\build\zlib\build> "%MSB%" /t:Build /p:Configuration=Release /p:Platform="x64" zlib.sln
C:\cms\build\zlib\build> "%MSB%" INSTALL.vcxproj
C:\cms\build> cd ..\..

C:\cms\build> [add full path]cmake -G "Visual Studio 15 2017 Win64" -DCMAKE_BUILD_TYPE=Release -DSTATICCOMPILE=ON -DZLIB_ROOT=C:\cms\build\zlib\myinstall -DBOOST_ROOT=C:\cms\build\boost_1_69_0_install ..
C:\cms\build> [add full path]cmake --build . --config Release

You now have the static binary under C:\cms\build\Release\cryptominisat5.exe

Error message:
LINK : fatal error LNK1181: cannot open input file 'cryptominisat5win.lib' [C:\cms\build\pycryptosat\python_interface.vcxproj]
CUSTOMBUILD : error : command 'E:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\VC\\Tools\\MSVC\\14.16.27023\\bin\\HostX86\\x64\\link.exe' failed with exit status 1181 [C:\cms\build\pycryptosat\python_interface.vcxproj]
```

**Comment by msoos** [comment]
Hey,

I have a weird error:
```
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin\x86_amd64\link.exe /ERRORREPORT:QUEUE /OUT:"C:\projects\cryptominisat\build\Release\cryptominisat5.exe" /INCREMENTAL:NO /NOLOGO /LIBPATH:C:/projects/cryptominisat/boost_1_59_0_install/lib /LIBPATH:C:/projects/cryptominisat/boost_1_59_0_install/lib/Release /LIBPATH:"C:\Tools\vcpkg\installed\x64-windows\lib" /LIBPATH:"C:\Tools\vcpkg\installed\x64-windows\lib\manual-link" ..\lib\Release\cryptominisat5win.lib ..\..\zlib\myinstall\lib\zlib.lib "..\..\boost_1_59_0_install\lib\boost_program_options-vc140-mt-1_59.lib" kernel32.lib user32.lib gdi32.lib winspool.lib shell32.lib ole32.lib oleaut32.lib uuid.lib comdlg32.lib advapi32.lib /MANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /PDB:"C:/projects/cryptominisat/build/Release/cryptominisat5.pdb" /SUBSYSTEM:CONSOLE /STACK:"1572864" /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:/projects/cryptominisat/build/lib/Release/cryptominisat5.lib" /MACHINE:X64  /machine:x64 /PDBCOMPRESS cryptominisat5.dir\Release\main.obj
  cryptominisat5.dir\Release\main_exe.obj
  cryptominisat5.dir\Release\signalcode.obj
LINK : fatal error LNK1104: cannot open file 'libboost_program_options-vc140-mt-1_59.lib' [C:\projects\cryptominisat\build\cmsat5-src\cryptominisat5.vcxproj]
Done Building Project "C:\projects\cryptominisat\build\cmsat5-src\cryptominisat5.vcxproj" (default targets) -- FAILED.
```

Do you know what could be the issue? Why the hell does this thing preprend "lib" to "boost_program_options-vc140-mt-1_59.lib"? I don't understand :( You see, `"..\..\boost_1_59_0_install\lib\boost_program_options-vc140-mt-1_59.lib"` does exist. But I don't understand why it prepends "lib" to it? Do you know what could be happening? Thanks!

**Comment by msoos** [comment]
Ah, build is at: https://ci.appveyor.com/project/msoos/cryptominisat/builds/21185330/job/k98dnvr8tb3pmeul

**Comment by a1880** [comment]
The "lib" prefix indicates that the library is for static linking.
libboost_program_options-vc140-mt-1_59.lib is the correct name, and the file exists.
boost_program_options-vc140-mt-1_59.lib also exists, but is for dynamic DLL linking.

**Comment by msoos** [comment]
Ah, that was a good hint actually! I will dig into this -- for whatever reason, it wanted to link it statically even though it's been asked to do it dynamically. Very strange. I'll dig into it a bit deeper. I finally have internet (yay! It only took 6 months in the capital one of the most developed countries in the world, Berlin! :D ) so I can now actually dig deeper :)

**Comment by a1880** [comment]
Great! I do like the statically linked binaries as they don't cause path and search problems. The library is much bigger than what finally ends up in the cms binary. The 5.6.6 cms binary is just 1.5mb big. That is quite small compared to solvers like Z3 with 10+mb. 

Welcome in Germany!

**Comment by a1880** [comment]
Have just seen the green AppVeyor building light. Congratulations!

**Comment by a1880** [comment]
I've just downloaded and tested the six artifacts using this script:

```
for /D %%d in (1 2 3 4 5 6 .) do call :doit %%d
goto xit

:doit
set CMS=%1\cryptominisat5.exe
if exist %CMS% (
  echo ..................................................................................
  echo testing %CMS%:
  %CMS% --verb=1 --threads=8 --polar=false --dumpresult=%1\result.txt s2x3x2_11.cnf.txt > %1\log.txt
  echo.
) else (
  echo %CMS% not found!
) 
goto :EOF

:xit
pause
```
The seventh contender was the executable from my own compilation.

All candidates ran well and solved my example within 7.3 and 9.0 seconds.
The binary sizes are all similar within +/- 100kb.

Everything looks fine!

**Comment by msoos** [comment]
Wow, thanks! Everything is fixed, thanks to time off from work :)

Happy holidays!

Mate
